pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+}
+
+func newCgroupV2Scanner() (utilizationScanner, bool) {
+	if _, err := os.Stat(memPathV2); err != nil {
+		return cgroupV2Scanner{}, false
+	}
+	if _, err := os.Stat(cpuPathV2); err != nil {
+		return cgroupV2Scanner{}, false
+	}
+	return cgroupV2Scanner{}, true
+}
+
+type cgroupV1Scanner struct {
+}
+
+func (s cgroupV1Scanner) String() string {
+	return "cgroup v1"
+}
+
+func (s cgroupV1Scanner) Scan() (float64, uint64, error) {
+	memR, err := readFileNoStat(memPathV1)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	// Summing RSS and cache for memory usage
+	// For reference, see section 5.5 in https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
+	var memUtil uint64
+	var numMemStats int
+	for {
+		var name string
+		var val uint64
+		if _, err := fmt.Fscanf(memR, "%s %d\n", &name, &val); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", memPathV1)
+		}
+
+		if name == "rss" || name == "cache" {
+			memUtil += val
+			numMemStats++
+			if numMemStats == 2 {
+				break
+			}
+		}
+	}
+	if numMemStats != 2 {
+		return 0, 0, fmt.Errorf("failed scanning %s", memPathV1)
+	}
+
+	// For reference, see https://www.kernel.org/doc/Documentation/cgroup-v1/cpuacct.txt
+	cpuR, err := readFileNoStat(cpuPathV1)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	// CPU time in nanoseconds
+	var cpuTime uint64
+	if _, err := fmt.Fscanf(cpuR, "%d", &cpuTime); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPathV1)
+	}
+
+	return float64(cpuTime) / 1e09, memUtil, nil
+}
+
+func newCgroupV1Scanner() (utilizationScanner, bool) {
+	if _, err := os.Stat(memPathV1); err != nil {
+		return cgroupV1Scanner{}, false
+	}
+	if _, err := os.Stat(cpuPathV1); err != nil {
+		return cgroupV1Scanner{}, false
+	}
+	return cgroupV1Scanner{}, true
+}
+
+type procfsScanner struct {
+	proc procfs.Proc
+}
+
+func (s procfsScanner) String() string {
+	return "/proc"
+}
+
+func (s procfsScanner) Scan() (float64, uint64, error) {
+	ps, err := s.proc.Stat()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process stats")
+	}
+
+	return ps.CPUTime(), uint64(ps.ResidentMemory()), nil
+}
+
+func newProcfsScanner() (utilizationScanner, bool) {
+	p, err := procfs.Self()
+	if err != nil {
+		return procfsScanner{}, false
+	}
+
+	return procfsScanner{
+		proc: p,
+	}, true
+}
+
+// UtilizationBasedLimiter is a Service offering limiting based on CPU and memory utilization.
+//
+// The respective CPU and memory utilization limits are configurable.
+type UtilizationBasedLimiter struct {
+	services.Service
+
+	logger             log.Logger
+	utilizationScanner utilizationScanner
+
+	// Memory limit in bytes
+	memoryLimit uint64
+	// CPU limit in cores
+	cpuLimit float64
+	// Last CPU time counter
+	lastCPUTime float64
+	// The time of the last update
+	lastUpdate     time.Time
+	movingAvg      *math.EwmaRate
```

Comment:
[nit] Just to clarify it.
```suggestion
	cpuMovingAvg      *math.EwmaRate
```

Reply:
Reviving this nit.